### PR TITLE
tidb_exporter

### PR DESCRIPTION
**Pull request by LinuxGit**
#### tidb_exporter
TiDB exporter for Prometheus. The tidb_exporter will expose metrics via HTTP on the /metrics endpoint. Currently only one tidb server address is supported. It is recommended to set load balancing IP. We'll support multiple tidb servers in the near future.

See the README.md file for more details.

### CODE REVIEW

File: tidb_exporter/main.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+	)
+
+	kingpin.Flag("tidb.address", "Address (host:port) of TiDB server.").Default("").StringVar(&opts.address)
+	kingpin.Flag("tidb.username", "TiDB user name.").Default("").StringVar(&opts.username)
+	kingpin.Flag("tidb.password", "TiDB user password.").Default("").StringVar(&opts.password)
+
+	kingpin.HelpFlag.Short('h')
+	kingpin.Parse()
+
+	log.SetLevelByString(*logLevel)
```
Comment (GregoryIan) [inline tidb_exporter/main.go]: do u want to check flags firstly?

File: tidb_exporter/main.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+	address  string
+	username string
+	password string
+}
+
+// NewExporter returns an initialized Exporter.
+func NewExporter(opts tidbOpts) (*Exporter, error) {
+	db, err := accessDatabase(opts.username, opts.password, opts.address, dbname)
+	if err != nil {
+		log.Errorf("access database '%s' error, %v", opts.address, err)
```
Comment (GregoryIan) [inline tidb_exporter/main.go]: u output it outside, dont need to output this error log again.   or u can use `juju/errors`'s `Annotationf` and `ErrorStack`

File: tidb_exporter/metrics.go [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+)
+
+const (
+	namespace = "tidb_exporter"
+)
+
+var (
+	queryErrorDesc = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "tidb", "query_error"),
+		"Whether an error occurred while sending query to tidb server.",
```
Comment (GregoryIan) [inline tidb_exporter/metrics.go]: occurs

File: tidb_exporter/main.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+	address  string
+	username string
+	password string
+}
+
+// NewExporter returns an initialized Exporter.
+func NewExporter(opts tidbOpts) (*Exporter, error) {
+	db, err := accessDatabase(opts.username, opts.password, opts.address, dbname)
+	if err != nil {
+		return nil, err
```
Comment (GregoryIan) [inline tidb_exporter/main.go]: errors.Trace(err)

File: tidb_exporter/main.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+	)
+
+	kingpin.Flag("tidb.address", "Address (host:port) of TiDB server.").Default("").StringVar(&opts.address)
+	kingpin.Flag("tidb.username", "TiDB user name.").Default("").StringVar(&opts.username)
+	kingpin.Flag("tidb.password", "TiDB user password.").Default("").StringVar(&opts.password)
+
+	kingpin.HelpFlag.Short('h')
+	kingpin.Parse()
+
+	if opts.address == "" {
```
Comment (GregoryIan) [inline tidb_exporter/main.go]: wrap these checking a function?

File: tidb_exporter/main.go [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+		syscall.SIGKILL,
+		syscall.SIGHUP,
+		syscall.SIGINT,
+		syscall.SIGTERM,
+		syscall.SIGQUIT)
+
+	go func() {
+		sig := <-sc
+		log.Infof("got signal [%d] to exit", sig)
+		exporter.db.Close()
```
Comment (GregoryIan) [inline tidb_exporter/main.go]: `if exporter.db != nil`

File: tidb_exporter/main.go [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+func checkParameters(opts tidbOpts) {
+	if opts.address == "" {
+		log.Fatalf("missing startup parameter: --tidb.address")
+	}
+
+	if opts.username == "" {
+		log.Fatalf("missing startup parameter: --tidb.username")
+	}
+
+	if opts.password == "" {
```
Comment (WangXiangUSTC) [inline tidb_exporter/main.go]: tidb's password can't be empty?

Comment (LinuxGit) [reply tidb_exporter/main.go]: In fact, checking opts.password is optional. I add it for security. I' ll remove it in next pr.

**Pull request merged**